### DISCUSSION TITLE
Generating Messages in non ROS environment (Docker)

### DISCUSSION THREAD

**Discussion by fxxholub** [noelex/rclnet#19] [Q&A]
Hi, i was wondering how to generate messages, when developing brand new package/node, which will be Docker based. I do not have any other ROS package nor do I have ROS installed on my dev machine, so I would like to know if there is a way to generate messages in this setup. Is there any way to generate at least Std messages?

There also i possibility that I did not understand how the message generating works. What is the input/source to the gen?

**Comment by noelex** [comment]
You just need to create a container using an image with `ros2cs` utility and ROS installed, and run `ros2cs` inside the container.

You can build an image using the following Dockerfile:
```Dockerfile
FROM mcr.microsoft.com/dotnet/sdk:6.0 AS installer
RUN dotnet tool install -g ros2cs

FROM ros:humble-ros-core
COPY --from=installer /root/.dotnet/tools /root/.dotnet/tools
RUN apt update && \
    apt install -y --no-install-recommends \
      dotnet-runtime-6.0 \
      ros-humble-nav2-msgs \
      ros-humble-tf2-msgs \
      ros-humble-cartographer-ros-msgs \
    && apt-get autoremove -y \
    && apt-get clean -y \
    && rm -rf /var/lib/apt/lists/*
ENV PATH=~/.dotnet/tools:${PATH}
```

And then run the image with your `ros2cs.spec` file and output directory mounted:
```
docker run -t --rm -v ros2cs.spec:/src/ros2cs.spec -v ./:/src ros2cs-runner ros2cs /src/ros2cs.spec
```
